presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/IndexedPriorityQueue.java

```diff
@@ -146,4 +147,59 @@ public long getGeneration()
             return generation;
         }
     }
+
+    private static class IndexedPriorityQueueIterator<E>
```

Comment:
we should make `peek` and `poll` also use this logic so that the traversal and cleanup logic is in a single position